### PR TITLE
[mxfp8 moe training] mxfp8 all to all

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Stacked PRs:
 * #2268
 * #2251
 * __->__#2250
 * #2249

--- --- ---

[mxfp8 moe training] mxfp8 all to all

## High level design

- The goal is to reduce all-to-all communication volume in expert parallelism, to achieve a speedup.
- Key idea: instead of waiting until directly before the grouped GEMM to quantize the inputs, we quantize a couple steps earlier, before the all2all, and stay in MXFP8 through the token shuffle, so the MXFP8 grouped GEMM receives pre-quantized input activations.
 
Screenshot 2026-02-04 at 11 00 34 AM

Forward flow:
                1. a2a dispatch forward: high-precision input, MXFP8 output
                2. permute forward: MXFP8 input, MXFP8 output
                3. MXFP8 grouped GEMM: MXFP8 input, high-precision output
                4. unpermute forward: high-precision input and output
                5. a2a combine forward: high-precision input and output

Backward flow:
                1. a2a combine backward: high-precision input, MXFP8 output
                2. unpermute backward: MXFP8 input, MXFP8 output
                3. MXFP8 grouped GEMM: MXFP8 input, high-precision output
                4. permute backward: high-precision input and output
                5. a2a dispatch backward: high-precision input and output

## Benchmarks
- +10% TPS on single node 8xB200 training with EP over NVLink domain
- +24% TPS on multi-node B200 cluster with IB inter-node networking (more bandwidth constrained -> bigger speedup)

## Limitations
- Not compatible with DeepEP
    - Recipe `"mxfp8"` IS compatible with DeepEP, but `"mxfp8_wgrad_with_hp"` (which uses MXFP8 all2all, is not)

## Testing
 
(test depends on the full PR stack)

```bash
CONFIG_FILE=/home/dev/torchtitan/torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml ./run_train.sh --metrics.log_freq=10 \
--training.steps=1500  \
--parallelism.data_parallel_shard_degree=4 \
--parallelism.expert_parallel_degree=4 \
--parallelism.tensor_parallel_degree=2 \
--parallelism.expert_tensor_parallel_degree=1 \
--training.seq_len=8192 \
--activation_checkpoint.mode=full \
--model.print_after_conversion \
--training.local_batch_size=16 \
--quantize.linear.mx.mxfp8_dim0_cast_kernel_choice="triton" --quantize.linear.mx.mxfp8_dim1_cast_kernel_choice="cuda" \
--quantize.grouped_mm.mx.fqns="experts" --quantize.grouped_mm.mx.recipe_name="mxfp8_wgrad_with_hp" \
--compile.enable --compile.components="model,loss" --debug.moe_force_load_balance \
--model.converters="quantize.grouped_mm.mx"
```

### CODE REVIEW

**Comment by danielvegamyhre** [comment]
@tianyu-l I will polish this up but any earlier thoughts on the high level design?

File: torchtitan/models/moe/utils.py [hunk 1]
```diff
@@ -95,3 +97,49 @@ def wrapper(
         return out
 
     return wrapper
+
+
+def get_a2a_splits(
```
Comment (tianyu-l) [inline torchtitan/models/moe/utils.py]: Given the current code organization, this function is a distributed util, so probably should be in expert_parallel.py.

File: torchtitan/config/job_config.py [hunk 2]
```diff
@@ -459,6 +459,54 @@ class Parallelism:
     https://github.com/deepseek-ai/DeepEP.
     """
 
+    expert_parallel_a2a_dispatch_fwd_precision: Literal["default", "mxfp8"] = "default"
```
Comment (tianyu-l) [inline torchtitan/config/job_config.py]: Is it not a net benefit to enable mxfp8 dispatch in fwd and mxfp8 combine in backward? If they are, they don't need to be configurable -- we can just enable them whenever EP and mxfp8 grouped mm is used.

Comment (danielvegamyhre) [reply torchtitan/config/job_config.py]: It is but there is an important limitation, which is that mxfp8 grouped mm recipe must be `mxfp8_wgrad_with_hp` (computes forward output and dgrad in mxfp8, computes wgrad in high precision). It is not compatible with `mxfp8` recipe yet (uses mxfp8 for all output/dgrad/wgrad). 

We could automatically enable it when the recipe is `mxfp8_wgrad_with_hp` perhaps?

Context on this requirement if you are curious:

This is because if inputs come in pre-quantized along dim0, in the backward pass `wgrad = grad_out_t @ input` where we need inputs quantized along dim1, we would need to dequant along dim0 then requant along dim1. This is definitely an option to do, especially if we have a fast fused kernel for this, however, there is some debate about the numerical implications of doing this (as it is not equivalent to doing dim1 quant on the original bf16 tensor). 

Therefore, for now we assert the recipe if `mxfp8_wgrad_with_recipe` pending (1) the fast dequant/requant kernel, and (2) numerics experiments.

Comment (tianyu-l) [reply torchtitan/config/job_config.py]: > It is but there is an important limitation, which is that mxfp8 grouped mm recipe must be mxfp8_wgrad_with_hp

Then we should couple it with `mxfp8_wgrad_with_hp`, unless you see reasons not to?

> This is because if inputs come in pre-quantized along dim0, in the backward pass wgrad = grad_out_t @ input where we need inputs quantized along dim1, we would need to dequant along dim0 then requant along dim1.

I get that wgrad_with_hp makes sense, but I didn't get what's its relationship to doing fwd combine / bwd dispatch of dgrad in high-precision.

Comment (danielvegamyhre) [reply torchtitan/config/job_config.py]: >Then we should couple it with mxfp8_wgrad_with_hp, unless you see reasons not to?

I agree this sounds good, the only blocker on that right now is that torch.compile is working in unit tests but not in torchtitan e2e training with these mxfp8 EP building blocks. cc @bdhirsh the nonstrict_trace dynamo decorator works in the torchao integration test w/ compile, but in torchtitan the same "tensor metadata mismatch" rears its head again. Will DM you about this.

Comment (bdhirsh) [reply torchtitan/config/job_config.py]: I made a diagram to describe why things are breaking:
image

torch.compile has a hard requirement today, where when you compile a given region of code, the forward outputs and their corresponding gradients (the backward inputs to that compiled region) must have the *same* subclass type.

You can see from this image that in the `mxfp8` codepath, both the `GroupedExperts.forward`, **and** the `MXFP8ExpertParallel._token_dispatch` violate this assumption.

That doesn't mean that we can't support this case. But it *does* mean that we are unable to compile these two blocks of code in isolation. Instead, our options are either:

(1) don't compile the token_dispatch or the `GroupedExperts.forward` at all
(2) compile a single graph, containing all of `_token_dispatch` + `GroupedExperts.forward` + `_token_combine`

I have a short diff for both 1 and 2, although I think there are a few things worth thinking more about.

diff for not compiling `GroupedExperts.forward`: https://gist.github.com/bdhirsh/970a671b84c35cc95a76f33657ca4d69

diff for compiling all 3 into a single graph: https://gist.github.com/bdhirsh/8b0ced2f6381b52eebd56fdff8e62093 This was a bit strange - it looks like when we try to compile `GroupedExperts`, dynamo ends up creating a separate graph for the `GroupedExperts.forward` and the module's forward pre-hooks / post-hooks. I "fixed" it by moving the `GroupedExperts` into an nn module wrapper and compiling that, but I imagine this isn't actually landable since it will e.g. muck up the state dict. I also saw some previous comments from @xmfan indicating that there were previous problems trying to compile `_token_dispatch/combine`: https://github.com/pytorch/torchtitan/issues/1940. I didn't understand the issue though.

Comment (bdhirsh) [reply torchtitan/config/job_config.py]: To me, the best option would be if we can tweak FSDP to handle multiple device meshes, so we can avoid having to FSDP-wrap the `experts` region (cc @weifengpy). Since I think that would solve this particular issue, and would also avoid a lot of the other issues that we've seen in titan around compiling MoE.

I'm not sure what the timeline for that is, so depending on @danielvegamyhre 's timeline, we might want to either just skip compiling the grouped_mm to unblock, or figure out how to land (2) above.

Comment (xmfan) [reply torchtitan/config/job_config.py]: Thanks for the clear diagram! Are there implementation changes we can do to make MXTensor <-> bf16 conversions to be moved into a different graph? Otherwise, I think the FSDP2 changes are necessary to compile a single graph with dispatch/experts/combine

Comment (bdhirsh) [reply torchtitan/config/job_config.py]: @danielvegamyhre can speak to that better than me, but I would imagine... probably not, unless you are willing to completely abandon the structure today of "_token_dispatch/combine" are module hooks that are logically separated from the grouped_mm forward call".

> Otherwise, I think the FSDP2 changes are necessary to compile a single graph with dispatch/experts/combine

I agree that this feels like an overall much cleaner state. One dumb question I have: I was able to get token_dispatch + grouped_mm + token_combine into a single graph with this change: https://gist.github.com/bdhirsh/8b0ced2f6381b52eebd56fdff8e62093. I agree it is ugly (and potentially not landable because it changes the state dict), but it seemed to work for me on Daniel's example. It sounds like there are other issues you ran into trying to compile the token_dispatch/combine - do you know what they are and what caused them?

Comment (xmfan) [reply torchtitan/config/job_config.py]: It seemed to be something with the unbacked shapes, I'm not sure if the error is still present. I believe the main consideration for module hook was to provide the same clean model definition for both ep and non-ep, so changing APIs seems okay as long as that we can ensure that

Comment (danielvegamyhre) [reply torchtitan/config/job_config.py]: > It is but there is an important limitation, which is that mxfp8 grouped mm recipe must be mxfp8_wgrad_with_hp

Then we should couple it with mxfp8_wgrad_with_hp, unless you see reasons not to?

@tianyu-l I went ahead made this change. One thing to note though is that this makes the `mxfp8_wgrad_with_hp` recipe incompatible with DeepEP. The standard `mxfp8` recipe (mxfp8 for all grouped gemms) is still compatible with DeepEP though, and we got great results with this combination recently (+41% TPS on DSV3 671B)

File: torchtitan/models/deepseek_v3/infra/parallelize.py [hunk 3]
```diff
@@ -120,6 +120,10 @@ def parallelize_deepseekv3(
     else:
         use_deepep = False
 
+    use_mxfp8_a2a = (
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/infra/parallelize.py:123]: Since we have DeepEP backend now enabled with this config https://github.com/pytorch/torchtitan/blob/main/torchtitan/config/job_config.py#L474
I think we should just add another backend mxfp8, and pass it to `apply_moe_ep_tp`, instead of using both `use_deepep` and `use_mxfp8_a2a`.

File: torchtitan/distributed/expert_parallel.py [hunk 4]
```diff
@@ -382,3 +468,49 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
             input_fn=self._token_dispatch,  # pyrefly: ignore [bad-argument-type]
             output_fn=self._token_combine,  # pyrefly: ignore [bad-argument-type]
         )
+
+
+def get_a2a_splits(
```
Comment (tianyu-l) [inline torchtitan/distributed/expert_parallel.py:473]: I now think it's better to make `_get_a2a_splits` a static method of ExpertParallel, so that subclasses of it can inherit.

File: torchtitan/distributed/expert_parallel.py [hunk 5]
```diff
@@ -192,6 +173,111 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
         )
 
 
+class MXFP8ExpertParallel(ExpertParallel):
```
Comment (tianyu-l) [inline torchtitan/distributed/expert_parallel.py:176]: Since it's related to mxfp8 only, let's put this in quantization/mx.py?

**Comment by weifengpy** [comment]
per-layer FSDP2 wrapping is landed. no more fsdp2 preventng tracing at moe boundary

next gap for per-layer compile is `mark_dynamic` at `_run_experts_grouped_mm_dynamic`: https://github.com/pytorch/torchtitan/pull/2741

**Comment by danielvegamyhre** [comment]
> per-layer FSDP2 wrapping is landed. no more fsdp2 preventng tracing at moe boundary
> 
> next gap for per-layer compile is `mark_dynamic` at `_run_experts_grouped_mm_dynamic`: #2741

amazing, thanks @weifengpy! will revisit this PR soon

**Comment by weifengpy** [comment]
> > next gap for per-layer compile is `mark_dynamic` at `_run_experts_grouped_mm_dynamic`: #2741
> amazing, thanks @weifengpy! will revisit this PR soon

@danielvegamyhre per-layer compile with moe works with this PR (no more mark_dynamic graph break): https://github.com/pytorch/torchtitan/pull/2741

I am waiting for another pytorch-side fix before landing it

**Comment by weifengpy** [comment]
> > per-layer FSDP2 wrapping is landed. no more fsdp2 preventng tracing at moe boundary
> > next gap for per-layer compile is `mark_dynamic` at `_run_experts_grouped_mm_dynamic`: #2741
> 
> amazing, thanks @weifengpy! will revisit this PR soon

Hi @danielvegamyhre , per-layer compile is landed https://github.com/pytorch/torchtitan/pull/2741

**Comment by danielvegamyhre** [comment]
@weifengpy awesome, thanks!
